# Assignment 3

## Instructions - Read this first!

This is an individual homework assignment. This means that:

- You may discuss the problems in this assignment with other students in this course and your instructor/TA, but YOUR WORK MUST BE YOUR OWN.
- Do not show other students code or your own work on this assignment.
- You may consult external references, but not actively receive help from individuals not involved in this course.
- Cite all references outside of the course you used, including conversations with other students which were helpful. (This helps us give credit where it is due!). All references must use a commonly accepted reference format, for example, APA or IEEE (or another citation style of your choice).

If any of these rules seem ambiguous, please check with with your instructor for help interpreting them.

We suggest completing this assignment using the provided notebook. Each question should be answered using a SQL query (or combination or SQL queries) unless the text indicates that you may (or should) do something else. You may submit your queries embedded in Python, using SQLAlchemy or the MySQL Connector, or as plain text in Markdown.

## When you submit your work

Your submission will be graded manually. To ensure that everything goes smoothly, please follow these instructions to prepare your notebook for submission to the D2L Dropbox for Assignment 3:

- Please remove any print statments used to test your work (you can comment them out)
- Please provide your solutions where asked; please do not alter any other parts of this notebook.
- If you need to add cells to test your code please move them to the end of the notebook before submission- or you may included your commented out answers and tests in the cells provided

## Introduction

In this assignment, you will continue to practice and extend your SQL skills, and compare your work to MongoDB.

We will be using two datasets from Open Data Calgary. The Census by Community 2019 dataset shows results of the 2019 Civic Census. This is a separate process from that used by Statistics Canada to produce the national census, some of which results we explored in the previous assignment. 

This census data counts dwelling units in the city and categorized the population within each unit, aggregating results by community. The original dataset has been partially cleaned for you, removing some columns which are not relevant to the assignment. However, given that there are still 75 columns remaining, you may want to use SQLAlchemy or another bulk load tool to handle the creation of the database table for this dataset, before altering the table as needed to change data types or creating indexes.

The Community Services dataset lists a number of amenities in the city with their address and community. This table has not been altered from the version provided by Open Data Calgary.

## Data cleaning and import

First, import the two CSVs (Community_Services.csv, Census_by_Community_2019.csv) into your own database. You may use what is available to you on datasciencedb or datasciencedb2. You may also create indexes and define keys if appropriate for the column(s) of your choice.

In the section below, you have the option to discuss any data cleaning and wrangling steps performed during this process. This is not a requirement and will not be assessed directly for grading; however, this may help to clarify to your reader exactly what was done, to make your work below more understandable.

In [2]:
#Data was loaded into my databse using MySQL WorkBench 

# Import libraries 
import pandas as pd
import sqlalchemy as sq

# Connect to databse
engine = sq.create_engine('mysql+mysqlconnector://kane_smith1:8ULM75GHS@datasciencedb2.ucalgary.ca/kane_smith1')

## Part A: Warm-up Questions (10 marks)

Answer the questions below, including any queries you used where necessary. Not all questions will require a SQL query for a correct response. You may wish to use as a source the references which are already provided as part of this notebook.

First, let's look at the Census by Community Data.

1. How many communities are included as part of this census? **(1 mark)**

2. There are 75 columns in the version of the dataset provided to you. While not all of them are equally interesting, briefly describe the following columns. **(3 marks)**

   - CLASS

   - SRG (include in your answer all posslbe values)

   - COMM_STUCTURE
   

3. What other columns, to your eye, look interesting? Provide an example of a guiding question which would take advantage of one of these columns. **(2 marks)**

**1.**

In [4]:
# Question 1 Query
query_table6 = pd.read_sql_query('''
SELECT COUNT(DISTINCT COMM_CODE) AS `Number of Unique Commuities`
FROM census
''', engine)
display(query_table6)



Number of Unique Commuities
0                          306

SRG
0  DEVELOPING
1   BUILT-OUT
2         N/A

There are 306 communities included in this census data


**2.** 

In [5]:
query_table6 = pd.read_sql_query('''
SELECT DISTINCT SRG 
FROM census
''', engine)
display(query_table6)

SRG
0  DEVELOPING
1   BUILT-OUT
2         N/A

a) <u>CLASS:</u> The identifier for what type of community an observation is. They can be either Residential, Residential Sub Area, Industrial or Major Park.

b) <u>SRG:</u> The status of housing development in the community. The values in the dataset are Developing, Built-out and N/A. The dataset source also says Non-residential is a possible value. (Census by Community 2019)

c) <u>COMM_STRICTURE:</u> An identifier of what demographic is present in the community. (examples of values are Inner city, employment and 1950s).

**3.** 

Columns that identify the occupancy of different building types (eg.CNV_OCCPD, DUP_OCCPD, APT_OCCPD etc.) as well as the total number of these building types. We could use this information to answer a guiding question related to occupancy rate for each community.

Next, let's look at the Community Services table. 

4. How many services in total are there? **(1 mark)**


5. List all possible types of community services listed in the table. How many are missing a community code? **(1 mark)**


6. Which column is the most useful as a key to be used in a join? **(1 mark)**


7. Which community has the highest number of available community services? Include the names of any community with this number of services. **(1 mark)**

**4.**

In [6]:
# Query for Question 4
query_table4 = pd.read_sql_query('''
SELECT COUNT(DISTINCT NAME) AS `Number of Unique Services`
FROM community_service
''', engine)
display(query_table4)

Number of Unique Services
0                        204

There are 204 services in the community services table

**5.**

In [38]:
# Query for Question 5
query_table6 = pd.read_sql_query('''
SELECT TYPE, SUM(CASE WHEN COMM_CODE='' THEN 1 ELSE 0 END) AS COUNT_NULL
FROM community_service
GROUP BY TYPE
''', engine)
display(query_table6)

TYPE  COUNT_NULL
0        Attraction         4.0
1  Community Centre         0.0
2             Court         0.0
3          Hospital         0.0
4           Library         0.0
5        PHS Clinic         0.0
6    Social Dev Ctr         0.0
7      Visitor Info         0.0

There are 8 types of community services (Community Centre, Attraction, Visitor Info, Court, Library, Hospital, PHS Clinic, and Social Dev Ctr). 4 are missing community codes. They are all attraction type "Attraction".

**6.**

The most useful column to do a join on is COMM_CODE. This is because it is also a column in the census table and each value is unique.

**7.**

In [9]:
#Query for Question 7
query_table7 = pd.read_sql_query('''
SELECT c.NAME, c.COMM_CODE, COUNT(s.NAME) AS COUNT_SERVICES
FROM census AS c
LEFT JOIN community_service AS s
ON c.COMM_CODE = s.COMM_CODE
GROUP BY c.NAME
ORDER BY COUNT_SERVICES DESC
''', engine)
display(query_table7)

NAME COMM_CODE  COUNT_SERVICES
0    DOWNTOWN COMMERCIAL CORE       DNC              24
1                    BELTLINE       BLN               6
2                 FOREST LAWN       FLN               5
3       UNIVERSITY OF CALGARY       UOC               4
4            HUNTINGTON HILLS       HUN               4
..                        ...       ...             ...
301               POINT MCKAY       POI               0
302                  SCARBORO       SCA               0
303                   MAYLAND       MLI               0
304              SKYLINE WEST       SKW               0
305                       10E       10E               0

[306 rows x 3 columns]

The community with the highest community services is DOWNTOWN COMMERCIAL CORE. They have 24 community services.

## Part B: SQL with multiple tables (13 marks)


What is the population of seniors (male and female, or other, aged 65 and over) in the community or communities with the most community services? **(2 marks)**

What is the population of children (male and female, or other, aged 0 to 14), in the community or communities with the most community services?  **(2 marks)**

What is the total population of seniors (male and female, or other, aged 65 and over) in all communities that contain nursing homes?  **(2 marks)**

What is the total number of residents, children (using the definition above) and seniors (using the definitions above) in communities which do not have any community services?  **(2 marks)**

Generate a list of the ten such communities with the greatest number of residents.  **(2 marks)**

Prepare a list of community services as follows: **(3 marks)**
- Each community service should list, for the community which they are in, the total number of residents, the total number of preschool children, and the number of dwellings.
- Community services with no community listed should use a meaningful placeholder in the result.
- Communities with no community services should be included in the results, with a meaningful placeholder where data for community services would ordinarily be included.

**1.** What is the population of seniors (male and female, or other, aged 65 and over) in the community or communities with the most community services

In [10]:
query_table9 = pd.read_sql_query('''
SELECT (c.MALE_65_74 + c.MALE_75) AS SENIOR_MALES, (c.FEM_65_74 + c.FEM_75) AS SENIOR_FEMALES, (c.OTHER_65_74 + c.OTHER_75) AS SENIOR_OTHER, (c.MALE_65_74 + c.MALE_75+c.FEM_65_74 + c.FEM_75+c.OTHER_65_74 + c.OTHER_75) AS TOTAL_SENIORS
FROM census AS c
WHERE NAME = "DOWNTOWN COMMERCIAL CORE"
''', engine)
display(query_table9)

SENIOR_MALES  SENIOR_FEMALES  SENIOR_OTHER  TOTAL_SENIORS
0         346.0           474.0             0          820.0

The community with the most services is Downtown Commerical Core. The total senior population there is 820.

**2.** What is the population of children (male and female, or other, aged 0 to 14), in the community or communities with the most community services? 

In [11]:
query_table9 = pd.read_sql_query('''
SELECT (c.MALE_0_4 + c.MALE_5_14) AS CHILD_MALES,(c.FEM_0_4 + c.FEM_5_14) AS CHILD_FEMALES,(c.OTHER_0_4 + c.OTHER_5_14) AS CHILD_OTHER, (c.MALE_0_4 + c.MALE_5_14+c.FEM_0_4 + c.FEM_5_14+c.OTHER_0_4 + c.OTHER_5_14) AS TOTAL_CHILDREN
FROM census AS c
WHERE NAME = "DOWNTOWN COMMERCIAL CORE"
''', engine)
display(query_table9)

CHILD_MALES  CHILD_FEMALES  CHILD_OTHER  TOTAL_CHILDREN
0        383.0          392.0           21           796.0

The community with the most services is Downtown Commerical Core. The total child population there is 796.

**3.** What is the total population of seniors (male and female, or other, aged 65 and over) in all communities that contain nursing homes? 

In [12]:
query_table9 = pd.read_sql_query('''
SELECT SUM(c.MALE_65_74 + c.MALE_75) AS TOTAL_SENIOR_MALES, SUM(c.FEM_65_74 + c.FEM_75) AS TOTAL_SENIOR_FEMALES,SUM(c.OTHER_65_74 + c.OTHER_75) AS TOTAL_SENIOR_OTHER,  SUM(c.MALE_65_74 + c.MALE_75+c.FEM_65_74 + c.FEM_75+c.OTHER_65_74 + c.OTHER_75) AS TOTAL_SENIORS
FROM census AS c
WHERE NURSING_HM > 0 
''', engine)
display(query_table9)

TOTAL_SENIOR_MALES  TOTAL_SENIOR_FEMALES  TOTAL_SENIOR_OTHER  TOTAL_SENIORS
0             25562.0               32704.0                27.0        58293.0

The population of seniors in communities with at least one nursing home is 58293.

**4.** What is the total number of residents, children (using the definition above) and seniors (using the definitions above) in communities which do not have any community services? 

In [37]:
query_table9 = pd.read_sql_query('''
SELECT COUNT(s.NAME) AS NUM_SERVICES, 
SUM((c.MALE_65_74)+(c.MALE_75)+(c.FEM_65_74)+(c.FEM_75)+(c.OTHER_65_74)+(c.OTHER_75)) AS TOTAL_SENIORS,
SUM((c.MALE_0_4)+(c.MALE_5_14)+(c.FEM_0_4)+(c.FEM_5_14)+(c.FEM_0_4)+(c.FEM_5_14)+(c.OTHER_0_4)+(c.OTHER_5_14)) AS TOTAL_CHILDREN, 
SUM(c.RES_CNT) AS TOTAL_RESIDENTS
FROM census AS c
LEFT JOIN community_service AS s
ON c.COMM_CODE = s.COMM_CODE
GROUP BY s.COMM_CODE
HAVING NUM_SERVICES = 0 
''', engine)
display(query_table9)

NUM_SERVICES  TOTAL_SENIORS  TOTAL_CHILDREN  TOTAL_RESIDENTS
0             0        46769.0        161705.0         500848.0

The total number of seniors, children and residents in communities where there are no services is 49769, 161705 and 500848.

**5.** Generate a list of the ten such communities with the greatest number of residents.

In [20]:
query_table9 = pd.read_sql_query('''
SELECT c.NAME, COUNT(s.NAME) AS NUM_SERVICES, 
(SUM(RES_CNT)) AS TOTAL_RESIDENTS
FROM census AS c
LEFT JOIN community_service AS s
ON c.COMM_CODE = s.COMM_CODE
GROUP BY c.NAME
HAVING NUM_SERVICES = 0
ORDER BY TOTAL_RESIDENTS DESC
LIMIT 10
''', engine)
display(query_table9)

NAME  NUM_SERVICES  TOTAL_RESIDENTS
0  PANORAMA HILLS             0          25710.0
1    SADDLE RIDGE             0          22321.0
2       EVERGREEN             0          21500.0
3        CRANSTON             0          19884.0
4        TARADALE             0          19026.0
5  MCKENZIE TOWNE             0          18283.0
6        EVANSTON             0          17685.0
7  COVENTRY HILLS             0          17667.0
8      AUBURN BAY             0          17607.0
9    NEW BRIGHTON             0          13103.0

**6.** 

In [23]:
query_table9 = pd.read_sql_query('''
SELECT c.NAME AS COMMUNITY_NAME, COALESCE(s.NAME, "No Community Services") AS SERVICE_NAME,  c.RES_CNT AS NUMBER_OF_RESIDENTS, PRSCH_CHLD AS PRESCHOOL_CHILDREN, c.DWELL_CNT AS NUMBER_OF_DWELLINGS
FROM census AS c
LEFT JOIN community_service AS s
ON c.COMM_CODE=s.COMM_CODE
GROUP BY c.NAME, s.NAME
''', engine)
display(query_table9)

COMMUNITY_NAME                                SERVICE_NAME  \
0                          01B                       No Community Services   
1                          01C                       No Community Services   
2                          01F                       No Community Services   
3                          01H                       No Community Services   
4                          01I                       No Community Services   
..                         ...                                         ...   
368  WINSTON HEIGHTS/MOUNTVIEW  Winston Heights/Mountview Community Centre   
369                WOLF WILLOW                       No Community Services   
370                   WOODBINE                       No Community Services   
371                  WOODLANDS                  Woodcreek Community Centre   
372                  YORKVILLE                       No Community Services   

     NUMBER_OF_RESIDENTS  PRESCHOOL_CHILDREN  NUMBER_OF_DWELLINGS  
0                    0.0                 0.0                  0.0  
1                    0.0                 0.0                  0.0  
2                    0.0                 0.0                  0.0  
3                    0.0                 0.0                  0.0  
4                    0.0                 0.0                  0.0  
..                   ...                 ...                  ...  
368               3635.0               247.0               1803.0  
369                  0.0                 0.0                 21.0  
370               8866.0               557.0               3372.0  
371               6002.0               438.0               2411.0  
372                 14.0                 0.0                118.0  

[373 rows x 5 columns]

## Part C: Evaluating your results  (7 marks)


**Question 1 (3 marks)**

The queries in Part B suggest that there are parts of Calgary which may be over-served by the numbers of community services available, and there are parts of Calgary which may be underserved by kinds of community services which are not available.

Before we continue with this line of inquiry, discuss two factors which may not be captured by either dataset.



Nominate one datset already available from Open Data Calgary which addresses one of your two factors. Remember to reference this using a well-accepted reference format.

**Factor 1: Public Transportation** Being able to easily travel to different areas in your city is very important for quality of life and getting opportunities in your career. Public transporation is crucial for people who do not have their own methods of transportation (eg. cars). Capturing this factor is important to see if a community is being under-served.

**Factor 2: Public Schooling** Having an education is critical to upward socioeconomic mobility. The number of public schools in a community is a good indicator of how well a government is serving the community in terms of education.

**Dataset: School Locations**  This dataset (School Locations) could be used to quantify the amount of public schools in a community in Calgary. (Dataset refrenced in the refrence section at the bottom of the notebook.)


**Question 2 (4 marks)**

Imagine that we have stored data relevant to each community in a separate JSON object, each of which has been added to a document in the same collection in MongoDB. Community Services have been added as list of JSON attributes to the relevant community.

For example, we might have
```
{
  Community:"Community Name", 
  ...
  Community_Services:[
      {
          "Community Service": "Community Service Name"
          
      },
      ...
      ]
  }
```
Pick one of the queries from Part B to discuss. Do you think it is more difficult to retrieve the information requested for this query from the pair of relational database tables provided to you, or from a MongoDB collection set up as described? Explain why or why not. 

**Query I chose:** What is the total number of residents, children (using the definition above) and seniors (using the definitions above) in communities which do not have any community services?

I think it would be more difficult to retrieve this information with a part of relational database tables. This is because with a MongoDB collection, we could simply filter the documents that have 0 community services and then sum the amount of children, seniors and residents.

With a relational database, we have to join two tables and do a calculation to see the number of services in each community, then filter to only have ones with 0 services. Then you would be able to sum children, seniors and residents.

The way information is organized in documents in MongoDB, it is easier to retreive, do calculations on and filter.

## Part E: Reflection (5 marks)


In 100 to 250 words, identify a concept you have found difficult or confusing from this assignment. Reflect on how your previous learning or experience helped you to understand this concept. Provide your reflection using markdown in the cell below.



A concept I found difficult in this assignment was joining tables and doing calculations on the values of these joined tables. This is because the code starts to get long as you add more and more keywords and variables. What I did to help with understanding the join and calculations that was required to answer the question was to ask myself "What variables do I need? What am I trying to calculate? (Do I need a COUNT() or a SUM()?), and do I need to group by any variables?" Also adding table aliasing help make joins simpler to read.

## References

Both datasets used in this Assignment are licensed under the Open Government License - City of Calgary.

Census by Community 2019 [online], 2022. Open Calgary. Available from: https://data.calgary.ca/Demographics/Census-by-Community-2019/rkfr-buzb [Accessed 26 Nov 2022].

Community Services [online], 2022. Open Calgary. Available from: https://data.calgary.ca/Services-and-Amenities/Community-Services/x34e-bcjz [Accessed 26 Nov 2022].

School Locations [online], 2022. Open Calgary. Available from: https://data.calgary.ca/Services-and-Amenities/School-Locations/fd9t-tdn2 [Accessed 01 Dec 2022].


In [ ]:
# Dispose of the engine object 
engine.dispose()